openedx_events/event_bus/avro/tests/test_deserializer.py

```diff
@@ -175,3 +176,40 @@ def test_deserialization_of_nested_optional_fields(self):
         nested_field = data_dict["data"].field_0
         self.assertIsInstance(nested_field, SimpleAttrsWithDefaults)
         self.assertEqual(nested_field, SimpleAttrsWithDefaults())
+
+    def test_deserialization_of_list_with_annotation(self):
+        """
+        Check that deserialization works as expected when list data is annotated.
+        """
+        LIST_SIGNAL = create_simple_signal({"list_input": List[int]})
+        initial_dict = {"list_input": [1, 3]}
+        deserializer = AvroSignalDeserializer(LIST_SIGNAL)
+        event_data = deserializer.from_dict(initial_dict)
+        expected_event_data = [1, 3]
+        test_data = event_data["list_input"]
+        self.assertIsInstance(test_data, list)
+        self.assertEqual(test_data, expected_event_data)
+
+    def test_deserialization_of_list_without_annotation(self):
+        """
+        Check that deserialization raises error when list data is not annotated.
+        """
+        SIGNAL = create_simple_signal({"list_input": List[int]})
+        LIST_SIGNAL = create_simple_signal({"list_input": List})
+        initial_dict = {"list_input": [1, 3]}
+        deserializer = AvroSignalDeserializer(SIGNAL)
+        deserializer.signal = LIST_SIGNAL
+        with self.assertRaises(TypeError):
+            deserializer.from_dict(initial_dict)
+
+    def test_deserialization_of_nested_list_fails(self):
+        """
+        Check that deserialization raises error when nested list data is passed.
+        """
+        SIGNAL = create_simple_signal({"list_input": List[int]})
+        LIST_SIGNAL = create_simple_signal({"list_input": List[List[int]]})
+        initial_dict = {"list_input": [[1, 3], [4, 5]]}
+        deserializer = AvroSignalDeserializer(SIGNAL)
+        deserializer.signal = LIST_SIGNAL
+        with self.assertRaises(TypeError):
+            deserializer.from_dict(initial_dict)
```

Comment:
Handling nested lists (with annotation) support again introduces complexity and I did not implement it as it is not required as of now. 

If we remove inner list type annotation, i.e. change the type to `List[list]` it works because the schema is becomes: `{'name': 'list_input', 'type': {'type': 'array', 'items': 'array'}}`

```python
    def test_deserialization_of_nested_list_works(self):
        SIGNAL = create_simple_signal({"list_input": List[list]})
        initial_dict = {"list_input": [[1, 3], [4, 5]]}
        deserializer = AvroSignalDeserializer(SIGNAL)
        data = deserializer.from_dict(initial_dict)
        self.assertEqual(data, initial_dict)
```

Reply:
Do we want List[list] to work? Doesn't that open us up to issues with something like `[[Object, 'a'], [1, "bananas"]]`?

Reply:
@rgraber Good point! I have updated it to allow only simple types like int, str, bool etc.